### ISSUE TITLE
Context should be merged, not replaced

### ISSUE THREAD

**Issue by MichalBryxi** [NullVoxPopuli/ember-statechart-component#196]
Assuming:

```js
const machine = createMachine(
  {
    id: 'foo',
    states: {
      ...
    },
    context: {
      foo: 'a',
    },
  }, {
    guards: {
      myGuard(context) {
        console.log({ context });
      },
  }
)
```

And:

```hbs
<Machine
  @context={{hash
    bar='b'
  }}
  as |state send|
>
</Machine>
```

I would expect to see:

```
{ context: { foo: 'a', bar: 'b' } }
```

But instead I see:

```
{ context: { bar: 'b' } }
```

So it seems to me like the `@context` replaces the `context` defined on the machine?

I think merging is preferred, because this: 
 
 1) Allows the developer to define the context of the machine correctly so that it can work correctly in the visualizer
 2) Allows the developer to define good default values
 3) Allows at the call site to override only those context variables that it cares about

**Comment by NullVoxPopuli** [comment]
That makes sense, i agree.

**Comment by nightire** [comment]
According to the xstate documents, `withContext` will not do a shallow merge of the origin context but instead will replace it.

It seems relatively easy to change this line:

https://github.com/NullVoxPopuli/ember-statechart-component/blob/58fe5d300811d3ffaa9bc4d7866bbbbc06430952/ember-statechart-component/src/-private/statechart-manager.ts#L51

to:

```ts
machine = machine.withContext(Object.assign({}, machine.context, context));
```

Does it good enough?

**Comment by NullVoxPopuli** [comment]
Perhaps!
Since withContext replaces, what does xstate docs recommend?
If we implement merging, how would someone replace?

**Comment by nightire** [comment]
> Perhaps! Since withContext replaces, what does xstate docs recommend? If we implement merging, how would someone replace?

Similar as I suggested above:

![xstate js org_docs_guides_machines html](https://user-images.githubusercontent.com/1104018/148231116-a1764a54-4ec6-42e3-9797-e1d9f3104175.png)

> If we implement merging, how would someone replace?

This is exactly what I concerned about. I think we should align with the xstate's behavior by provide an argument to allow context merging:

```hbs
<Machine @context={{hash ...}} @mergeContext={{true}} />
```

**Comment by nightire** [comment]
Both `@xstate/react`, `@xstate/vue` and `@xstate/svelte` merges context by default:

https://github.com/statelyai/xstate/blob/c8b1802451e6e787a8c516dee81aa4e017a5906e/packages/xstate-react/src/useInterpret.ts#L82-L93

https://github.com/statelyai/xstate/blob/c8b1802451e6e787a8c516dee81aa4e017a5906e/packages/xstate-svelte/src/useMachine.ts#L58-L61

https://github.com/statelyai/xstate/blob/a26897174a85672e70cba788a717c95b53b26235/packages/xstate-vue/src/useInterpret.ts#L70-L73

I'm not sure about the extra argument idea, maybe it's better to merge the context by default, thus keeping the API simple and clean. If replacing is required by others then we can provide an optional argument later.

**Comment by NullVoxPopuli** [comment]
> maybe it's better to merge the context by default, thus keeping the API simple and clean. If replacing is required by others then we can provide an optional argument later.


Excellent, let's do this, and then release as a breaking change

**Comment by nightire** [comment]
Okay, I will do this: https://github.com/NullVoxPopuli/ember-statechart-component/pull/204

**Comment by NullVoxPopuli** [comment]
This was released in 4.0.0
